### ISSUE TITLE
runtime: windows callback on non-Go thread fails

### ISSUE THREAD

**Issue by alexbrainman** [golang/go#6751]
<pre>This test

diff --git a/src/pkg/runtime/syscall_windows_test.go
b/src/pkg/runtime/syscall_windows_test.go
--- a/src/pkg/runtime/syscall_windows_test.go
+++ b/src/pkg/runtime/syscall_windows_test.go
@@ -240,5 +240,35 @@
 }
 
 func TestCallbackInAnotherThread(t *testing.T) {
-   // TODO: test a function which calls back in another thread: QueueUserAPC() or
CreateThread()
+   d := GetDLL(t, &quot;kernel32.dll&quot;)
+
+   f := func(p uintptr) uintptr {
+       return p
+   }
+   r, _, err := d.Proc(&quot;CreateThread&quot;).Call(0, 0, syscall.NewCallback(f), 123,
0, 0)
+   if r == 0 {
+       t.Fatalf(&quot;CreateThread failed: %v&quot;, err)
+   }
+   h := syscall.Handle(r)
+   defer syscall.CloseHandle(h)
+
+   switch s, err := syscall.WaitForSingleObject(h, 100); s {
+   case syscall.WAIT_OBJECT_0:
+       break
+   case syscall.WAIT_TIMEOUT:
+       t.Fatal(&quot;timeout waiting for thread to exit&quot;)
+   case syscall.WAIT_FAILED:
+       t.Fatalf(&quot;WaitForSingleObject failed: %v&quot;, err)
+   default:
+       t.Fatalf(&quot;WaitForSingleObject returns unexpected value %v&quot;, s)
+   }
+
+   var ec uint32
+   r, _, err = d.Proc(&quot;GetExitCodeThread&quot;).Call(uintptr(h),
uintptr(unsafe.Pointer(&amp;ec)))
+   if r == 0 {
+       t.Fatalf(&quot;GetExitCodeThread failed: %v&quot;, err)
+   }
+   if ec != 123 {
+       t.Fatalf(&quot;expected 123, but got %d&quot;, ec)
+   }
 }

fails with

C:\&gt;u:\test -test.v -test.run=Another
=== RUN TestCallbackInAnotherThread
--- FAIL: TestCallbackInAnotherThread (0.11 seconds)
        syscall_windows_test.go:259: timeout waiting for thread to exit
FAIL

hg id is 39c724dd7f25

Alex

PS: for after Go 1.2</pre>

**Comment by stevenh** [comment]
Simple test case:
```package main

import (
    //"C" // Uncomment this and it works
    "fmt"
    "syscall"
)

func ThreadProc(p uintptr) uintptr {
    fmt.Println("FOO")
    return 0
}

func main() {
    modkernel32 := syscall.MustLoadDLL("kernel32.dll")
    procCreateThread := modkernel32.MustFindProc("CreateThread")
    r1, _, _ := procCreateThread.Call(0, 0, syscall.NewCallback(ThreadProc), 0, 0, 0)
    h := syscall.Handle(r1)
    syscall.WaitForSingleObject(h, syscall.INFINITE)
    syscall.CloseHandle(h)
}
```

In this case syscall.NewCallback(..) appears to be key, so possibly it should have a warning about this issue, but ideally we should fix this.

@rsc suggestion made in #9240 of always enabling "C" under Windows seems reasonable, if its an option.

**Comment by alexbrainman** [comment]
> @rsc suggestion made in #9240 of always enabling "C" under Windows seems reasonable, if its an option.

Importing "C" package makes your program use Cgo. So your approach does not seems reasonable to me personally.

Alex

**Comment by billziss-gh** [comment]
Are there any plans to fix this problem?

My specific need is the following: I have [cgofuse](https://github.com/billziss-gh/cgofuse), a cross-platform FUSE (Filesystem in Userspace) package based on cgo. I am hoping to eliminate the need for cgo on Windows, by using `syscall.DLL`, `syscall.Proc` and `syscall.NewCallbackCDecl`, but I am blocked by this issue.

This is to support the [go-ipfs](https://github.com/ipfs/go-ipfs) project, which would rather avoid the use of a C compiler on Windows.

More information here: https://github.com/ipfs/go-ipfs/issues/5003#issuecomment-390868198

**Comment by djdv** [comment]
It may be worth mentioning that my original intention was to write a dynamic binding for [winfsp](https://github.com/billziss-gh/winfsp) separate from cgofuse. This issue prevents either effort from progressing.

**Comment by zhangyoufu** [comment]
@djdv try `import _ "runtime/cgo"` as a workaround

**Comment by billziss-gh** [comment]
@zhangyoufu 

```
> set CGO_ENABLED=0

> go version
go version go1.10 windows/amd64

> go build ./examples/memfs
# github.com/billziss-gh/cgofuse/examples/memfs
_cgo_init: relocation target x_cgo_init not defined
_cgo_notify_runtime_init_done: relocation target x_cgo_notify_runtime_init_done not defined
_cgo_thread_start: relocation target x_cgo_thread_start not defined
runtime.cgo_yield: relocation target _cgo_yield not defined
_cgo_init: undefined: "x_cgo_init"
_cgo_notify_runtime_init_done: undefined: "x_cgo_notify_runtime_init_done"
_cgo_thread_start: undefined: "x_cgo_thread_start"
runtime.cgo_yield: undefined: "_cgo_yield"
```

**Comment by alexbrainman** [comment]
> Are there any plans to fix this problem?

There is no one working on this. If you want this fixed, I suggest you try and fix it yourself. This issue has a test, so all you need to do is adjust runtime package to make test pass. :-) If  you need any help, just ask here. We will try and answer any questions you might have.

Alex

**Comment by billziss-gh** [comment]
@alexbrainman thanks. It is unlikely I will have the time to look into this for a few days, but I am hoping to find some time for experimentation in the near future.

**Comment by billziss-gh** [comment]
Actually this problem appears to happen with cgo as well.

Consider the following go program:

```go
package main

func main() {
	select {}
}
```

Run it to find that `checkdead` works:

```console
$ go run deadlock.go 
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [select (no cases)]:
main.main()
	/Users/billziss/Projects/go/src/deadlock.go:4 +0x20
exit status 2
```

Now consider the following cgo program:

```go
package main

import "C"

func main() {
	select {}
}
```

Run it to find that `checkdead` does not work and the program hangs:

```console
$ go run deadlock.go 
^Csignal: interrupt
```

I believe that the problem lies with `checkdead`, which should be improved to handle the extra M in cgo and the patch above.

**Comment by billziss-gh** [comment]
The following patch passes all tests on Windows and macOS. (On Windows this excludes the 2 tests `TestBreakpoint` and `TestAbort` which freeze my Windows VM, even when the patch is **not** applied.)

```diff
diff --git a/src/runtime/proc.go b/src/runtime/proc.go
index ba76f7c3e7..1647044f36 100644
--- a/src/runtime/proc.go
+++ b/src/runtime/proc.go
@@ -1287,7 +1287,7 @@ func mstart1() {
 //go:yeswritebarrierrec
 func mstartm0() {
 	// Create an extra M for callbacks on threads not created by Go.
-	if iscgo && !cgoHasExtraM {
+	if (iscgo || GOOS == "windows") && !cgoHasExtraM {
 		cgoHasExtraM = true
 		newextram()
 	}
@@ -1612,7 +1612,7 @@ func allocm(_p_ *p, fn func()) *m {
 // put the m back on the list.
 //go:nosplit
 func needm(x byte) {
-	if iscgo && !cgoHasExtraM {
+	if (iscgo || GOOS == "windows") && !cgoHasExtraM {
 		// Can happen if C/C++ code calls Go from a global ctor.
 		// Can not throw, because scheduler is not initialized yet.
 		write(2, unsafe.Pointer(&earlycgocallback[0]), int32(len(earlycgocallback)))
@@ -4209,8 +4209,13 @@ func checkdead() {
 		return
 	}
 
+	var run0 int32
+	if !iscgo && cgoHasExtraM {
+		run0 = 1
+	}
+
 	run := mcount() - sched.nmidle - sched.nmidlelocked - sched.nmsys
-	if run > 0 {
+	if run > run0 {
 		return
 	}
 	if run < 0 {
diff --git a/src/runtime/syscall_windows_test.go b/src/runtime/syscall_windows_test.go
index dfde12a211..fd5ecca03d 100644
--- a/src/runtime/syscall_windows_test.go
+++ b/src/runtime/syscall_windows_test.go
@@ -251,7 +251,34 @@ func TestBlockingCallback(t *testing.T) {
 }
 
 func TestCallbackInAnotherThread(t *testing.T) {
-	// TODO: test a function which calls back in another thread: QueueUserAPC() or CreateThread()
+	d := GetDLL(t, "kernel32.dll")
+	f := func(p uintptr) uintptr {
+		return p
+	}
+	r, _, err := d.Proc("CreateThread").Call(0, 0, syscall.NewCallback(f), 123, 0, 0)
+	if r == 0 {
+		t.Fatalf("CreateThread failed: %v", err)
+	}
+	h := syscall.Handle(r)
+	defer syscall.CloseHandle(h)
+	switch s, err := syscall.WaitForSingleObject(h, 100); s {
+	case syscall.WAIT_OBJECT_0:
+		break
+	case syscall.WAIT_TIMEOUT:
+		t.Fatal("timeout waiting for thread to exit")
+	case syscall.WAIT_FAILED:
+		t.Fatalf("WaitForSingleObject failed: %v", err)
+	default:
+		t.Fatalf("WaitForSingleObject returns unexpected value %v", s)
+	}
+	var ec uint32
+	r, _, err = d.Proc("GetExitCodeThread").Call(uintptr(h), uintptr(unsafe.Pointer(&ec)))
+	if r == 0 {
+		t.Fatalf("GetExitCodeThread failed: %v", err)
+	}
+	if ec != 123 {
+		t.Fatalf("expected 123, but got %d", ec)
+	}
 }
 
 type cbDLLFunc int // int determines number of callback parameters
```

**Comment by alexbrainman** [comment]
> Despite my earlier statement about needing a few days to look into this I decided to have a look

Did not take you long to come up with a solution. :-)

> Running all.bat hangs the whole OS(!)

I have never seen anything like that. If you care, you can create new issue and someone will investigate.

> I have traced this problem to the following 2 tests: TestBreakpoint and TestAbort

Normally hanged test are debugged by adding -timeout parameter. -timeout forces test to crash, and crash prints stack trace, and you can see where test hangs. But your test hangs whole OS, so I don't see how my advise is useful here.

> I believe that the problem lies with checkdead, which should be improved to handle the extra M in cgo and the patch above.

I do not know, if your program should hang or not. So I created #25538 to confirm.

> The following patch passes all tests on Windows

Yes this works for me too. On Windows 10 here.

I am not familiar with scheduler, so I cannot confirm your patch is correct or not. But if you send your change for review https://golang.org/doc/contribute.html someone will check your code.

Thank you.

Alex

**Comment by billziss-gh** [comment]
> I am not familiar with scheduler, so I cannot confirm your patch is correct or not. But if you send your change for review https://golang.org/doc/contribute.html someone will check your code.

Thanks for the pointer. The document mentions Gerrit, which I am not familiar with, but it looks like I can now submit the patch as a GitHub PR as well. The Google CLA looks fairly benign, so I an inclined to sign it.

Re: the patch itself. A large portion of this patch includes your test @alexbrainman which I would not be able to submit under the Google CLA (I would be violating term 5 or have to follow the process in term 7, which I do not understand). So I can only submit my changes against `runtime/proc.go`.

**Comment by billziss-gh** [comment]
I have now submitted PR # 25575, which includes the changes in `runtime/proc.go`, but not the test in `TestCallbackInAnotherThread`. (As explained above I cannot submit the test as it is not my work.)

**Comment by alexbrainman** [comment]
> it looks like I can now submit the patch as a GitHub PR as well.

Yes that should work. Unfortunately your PR #25575 seems stuck - you created and signed CLA 4 hours ago, and Gobot should have created Gerrit CL by now. We'll give it little more time, and maybe ping for help. It is not good that it is weekend now, perhaps we'll have to wait for next week.

> large portion of this patch includes your test @alexbrainman which I would not be able to submit under the Google CLA

I contributed to Go before, so I have signed CLA, and you could take my code as part of your change. But I created https://go-review.googlesource.com/c/go/+/114755 in case we want two separate changes. I don't care how, as long as your change goes in with the test.

> I have now submitted PR # 25575, which includes the changes in runtime/proc.go

Thank you for fixing this. It have very little free time nowdays.

Alex

**Comment by billziss-gh** [comment]
> Unfortunately your PR #25575 seems stuck...

I see. Obviously I am unfamiliar with the process, so I did not know what to expect. Perhaps the "gobot" needs a kick? :)

> I contributed to Go before, so I have signed CLA...

You mean to tell me that you do not work for Google, but somehow you have 710+ commits on this project? Respect!

> and you could take my code as part of your change. But I created https://go-review.googlesource.com/c/go/+/114755 in case we want two separate changes. I don't care how, as long as your change goes in with the test.

I would prefer it if we kept the changes separate. The CLA states in term 5 that: "You represent that each of Your Contributions is Your original creation...". I find it inadvisable to knowingly violate the CLA on the same day I signed it :)

> Thank you for fixing this. It have very little free time nowdays.

You are welcome. I very much understand the no-time argument.

**Comment by alexbrainman** [comment]
> Perhaps the "gobot" needs a kick? :)

Perhaps it does need a kick. I pinged Brad on your PR, maybe he will help. It is getting late Friday in West coast of USA, so it might be a while when someone will help - Go Team members are real people. But we will get your change reviewed.

> You mean to tell me that you do not work for Google,

No, I do not work for Google.

Alex

**Comment by tohosys** [comment]
Dear alexbrainman 
I got a same problem like this;
- https://groups.google.com/forum/#!topic/golang-nuts/h91eDx27vrQ

I tried your old below sample but same result.
- https://github.com/golang/go/issues/6751#issuecomment-66087949
   - Attachments:
      - multithread.go (1138 bytes)
      - multithreaded.zip (6031 bytes)

Could you please give me some solution ?
Thank you very much~

Best Regards,
JayKim

**Comment by alexbrainman** [comment]
> I got a same problem like this;

Same problem as #6751 ? As you can see, the issue is still opened, so the problem still remains. But @billziss-gh proposed solution - you can try his solution at https://golang.org/cl/114802.

Alex
